memory: add uv_replace_allocator

beevik opened the pull request:
With uv_replace_allocator, it's possible to override the default
memory allocator's malloc and free calls with functions of the user's
choosing. This allows libuv to interoperate with projects requiring a
custom memory allocator.

Internally, all calls to malloc and free have been replaced with
uv__malloc and uv__free, respectively. The uv__malloc and uv__free
functions call malloc and free unless they have been overridden by a
previous call to uv_replace_allocator.

As part of this change, the special aligned memory allocations
performed in src/win/fs-event.c has been replaced with standard
allocations. The 4-byte alignment being requested in this file was
unnecessary, since standard allocators already guarantee at least an
8-byte alignment.

beevik commented:
I apologize for making another pull-request out of this, but my git-fu is weak, and I was unable to bring [PR#178](https://github.com/libuv/libuv/pull/178) into proper shape.

This change contains all the changes from the previous pull request, which I have closed.

saghul commented on docs/src/misc.rst:
```diff
@@ -114,6 +114,12 @@ API
     Returns the libuv version number as a string. For non-release versions
     "-pre" is appended, so the version number could be "1.2.3-pre".
 
+.. c:function:: void uv_replace_allocator(uv_malloc_func malloc_func, uv_free_func free_func)
```
Please add typedefs for uv_malloc_func and uv_free_func where datatypes are defined.

saghul replied:
int uv_replace_allocator

saghul commented on docs/src/misc.rst:
```diff
@@ -114,6 +114,12 @@ API
+
+    Override the use of the standard library's malloc and free functions for
+    memory allocation. If used, this function must be called before any
+    other libuv function is called. It may not be called more than once.
```
Mention that it returns UV_EINVAL in that case, or 0 on success.

saghul commented on include/uv.h:
```diff
@@ -244,6 +244,15 @@ typedef enum {
+typedef void (*uv_free_func)(void* ptr);
+
+/*
+ * If used, uv_replace_allocator must be called before any other uv 
```
This is already mentioned in the docs and doesn't really add much here, please remove it,

saghul commented:
Address the comments to the documentation and LGTM. Good work!

beevik commented:
I've updated the documentation and uv.h.  If you think it looks good, I'll squash the 2 commits (and rebase if necessary).

Thanks for your patience and help.

saghul commented:
LGTM. @bnoordhuis, @piscisaureus ?

saghul commented:
Oh, one last thing: add a ".. versionadded:: 1.5.0" thing please.

bnoordhuis commented on include/uv-win.h:
```diff
@@ -45,8 +45,8 @@ typedef struct pollfd {
-#  define _malloca(size) malloc(size)
-#  define _freea(ptr) free(ptr)
+#  define _malloca(size) uv__malloc(size)
+#  define _freea(ptr) uv__free(ptr)
```
Does this actually work?  uv-win.h is a semi-public header, it gets included by the downstream project, whereas `uv__malloc()` and `uv__free()` are internal.

For that matter, I wonder why these defines are in uv-win.h in the first place.  Maybe this block should just be removed?  @piscisaureus?

beevik replied:
Oh wow, didn't realize that was a public header.

It looks like the only code inside libuv that's using _malloca is a single function in src/win/process.c.  So this could either be moved to process.c, or the code could just always do a heap allocation.  (I doubt the stack allocation is saving much as far as this optimization goes.)

bnoordhuis replied:
I agree, I think it can be scrapped.  It'd be nice to have confirmation from @piscisaureus though.

saghul replied:
FWIW I'm +1 to always doing a heap allocation.
On Feb 27, 2015 7:00 PM, "Ben Noordhuis" <notifications@github.com> wrote:

> In include/uv-win.h
> <https://github.com/libuv/libuv/pull/231#discussion_r25524815>:
>
> > @@ -45,8 +45,8 @@ typedef struct pollfd {
> >
> >  #ifndef _malloca
> >  # if defined(_DEBUG)
> > -#  define _malloca(size) malloc(size)
> > -#  define _freea(ptr) free(ptr)
> > +#  define _malloca(size) uv__malloc(size)
> > +#  define _freea(ptr) uv__free(ptr)
>
> I agree, I think it can be scrapped. It'd be nice to have confirmation
> from @piscisaureus <https://github.com/piscisaureus> though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/libuv/libuv/pull/231/files#r25524815>.
>

saghul replied:
@piscisaureus quick comment here? please :-)

bnoordhuis commented on include/uv.h:
```diff
@@ -244,6 +244,11 @@ typedef enum {
+typedef void* (*uv_malloc_func)(size_t size);
+typedef void (*uv_free_func)(void* ptr);
+
+UV_EXTERN int uv_replace_allocator(uv_malloc_func malloc_func, uv_free_func free_func);
```
Style nit: long line, please put the second argument on a new line and make it line up with the first.  (Or I can do that before I land it.)

bnoordhuis commented:
LGTM with a question and a style nit.  I noticed there is some trailing whitespace in the documentation.

saghul commented:
@beevik can you please replace the use of malloca/freea by always doing a heap allocation? Then squash your commits into one and I'll land it.

Thanks a lot for working on this!

beevik commented:
Ok, rebased and squashed!

I also converted pipe.c to use malloc/free instead of _malloca/_freea.  (I removed _malloca and _freea #defines from uv-win.h).

beevik commented:
By the way, I imagine a number of outstanding forks are still calling free/malloc, so I'll keep an eye on the repo for a while to watch for code that isn't calling uv__malloc/uv__free.

saghul commented:
LGTM. @bnoordhuis, can you please have one last look?

bnoordhuis commented:
Rubber-stamp LGTM.  Thanks for doing this, @beevik.

saghul commented:
Thanks a lot @beevik! Landed in :sparkles: c272f1f :sparkles:
